File: Luogu/p2000+/p2000-p2999/p2916/p2916.cpp
#include<bits/stdc++.h>
using namespace std;
const int N = 10010, M = 100010;
int n, m;
struct Edge
{
	int a, b, w;
	bool operator< (const Edge &t)const
	{
		return w < t.w;
	}
}e[M];
int p[N];
int find(int x)
{
	if(p[x] != x) p[x] = find(p[x]);
	return p[x];
}
int reich[N];
int main()
{
	scanf("%d%d", &n, &m);
	{
		int minn = 0x3f3f3f3f;
		memset(e, 0, sizeof(e));
		for(int i = 1; i <= n; i++) p[i] = i;
		for(int i = 1; i <= n; i++)scanf("%d", &reich[i]), minn = min(minn, reich[i]);
		for(int i = 0; i < m; i++)
		{
			int a, b, w;
			scanf("%d%d%d", &a, &b, &w);
			e[i] = { a, b, w * 2 + reich[a] + reich[b] };
		}
		sort(e, e + m);
		int res = 0;
		for(int i = 0; i < m; i++)
		{
			int a = find(e[i].a), b = find(e[i].b), w = e[i].w;
			if(a != b)p[a] = b, res += e[i].w;
		}
		cout << res + minn << endl;
	}
	return 0;
}

File: Luogu/p2000-p2999/p2916/p2916.java
import java.util.Arrays;
import java.lang.Math;
import java.util.Scanner;

public class Main{
	static int[] p=new int[10010];
	public static void main(String[] args) {
		Scanner scan= new Scanner(System.in);
		int n=scan.nextInt();
		int m=scan.nextInt();
		int[] num=new int[n+1];
		Edge[] e=new Edge[m];
		int minn=0x3f3f3f3f;
		for(int i=1;i<=n;i++)p[i]=i;
		for(int i=1;i<=n;i++){
			num[i]=scan.nextInt();
			minn=Math.min(minn,num[i]);
		}
		for(int i=0;i<m;i++){
			int a=scan.nextInt();
			int b=scan.nextInt();
			int w=scan.nextInt();
			e[i]=new Edge(a,b,w*2+num[a]+num[b]);
		}
		Arrays.sort(e);
		int ans=0;
		for(int i=0;i<m;i++){
			int a=e[i].a,b=e[i].b,w=e[i].w;
			int pa=find(a),pb=find(b);
			if(pa!=pb){
				p[pa]=pb;
				ans+=w;
			}
		}
		System.out.printf("%d\n",ans+minn);
		scan.close();
	}

	static class Edge implements Comparable<Edge>{
		int a,b,w;
		public int compareTo(Edge a){
			return this.w-a.w;
		}
		public Edge(int _a,int _b,int _w){
			this.a=_a;
			this.b=_b;
			this.w=_w;
		}
	}

	public static int find(int x){
		if(p[x]!=x)p[x]=find(p[x]);
		return p[x];
	}
}